src/evm/protocol/uniswap_v4/hooks/hook_handler_creator.rs

```diff
@@ -49,9 +68,82 @@ pub struct GenericVMHookHandlerCreator;
 impl HookHandlerCreator for GenericVMHookHandlerCreator {
     fn instantiate_hook_handler(
         &self,
-        _params: HookCreationParams,
+        params: HookCreationParams<'_>,
     ) -> Result<Box<dyn HookHandler>, InvalidSnapshotError> {
-        todo!()
+        // TODO double check how the hook address and bytecode attributes are actually called
+        let hook_address_bytes = params
+            .attributes
+            .get("hook_address")
+            .ok_or_else(|| InvalidSnapshotError::MissingAttribute("hook_address".to_string()))?;
+
+        let hook_address = Address::from_slice(&hook_address_bytes.0);
+
+        let hook_bytecode_bytes = params
+            .attributes
+            .get("hook_bytecode")
+            .ok_or_else(|| InvalidSnapshotError::MissingAttribute("hook_bytecode".to_string()))?;
+
+        let bytecode =
+            Bytecode::new_raw(alloy::primitives::Bytes::from(hook_bytecode_bytes.0.clone()));
+
+        let block_header = BlockHeader {
+            number: params.block.number,
+            hash: params.block.hash,
+            timestamp: params.block.timestamp,
+        };
+
+        let db = SimulationDB::new(get_client(None), get_runtime(), Some(block_header));
+
+        let engine = create_engine(db, true).map_err(|e| {
+            InvalidSnapshotError::VMError(SimulationError::FatalError(format!(
+                "Failed to create engine: {e:?}"
+            )))
+        })?;
+
+        // TODO what is the difference between params.account_balances and params.balances?
+        for account_bytes in params.account_balances.keys() {
+            let account_address = Address::from_slice(&account_bytes.0);
+
+            let account_balance = U256::from(1000000000000000000u64); // 1 ETH default
+
+            engine.state.init_account(
+                account_address,
+                AccountInfo {
+                    balance: account_balance,
+                    nonce: 0,
+                    code_hash: B256::ZERO,
+                    code: None, // No code for regular accounts
+                },
+                None,
+                false,
+            );
+        }
+
+        // Initialize all token contracts
+        for token_address_bytes in params.all_tokens.keys() {
+            let token_address = Address::from_slice(&token_address_bytes.0);
+
+            // Deploy ERC20 contract for this token
+            let erc20_bytecode = Bytecode::new_raw(alloy::primitives::Bytes::from(ERC20_BYTECODE));
+            let code_hash = B256::from(keccak256(erc20_bytecode.clone().bytes()));
+
+            engine.state.init_account(
+                token_address,
+                AccountInfo {
+                    balance: U256::ZERO, // Token contracts have zero ETH balance
+                    nonce: 1,
+                    code_hash,
+                    code: Some(erc20_bytecode),
+                },
+                None,
+                false,
+            );
```

Comment:
why are you doing this?? 🤔

Reply:
I'm not so sure that this is needed and we talked about this the other day remember? 🤔 and we said: let's not do it until we are sure we need it.
and if this is needed I think this should be done inside the GenericVMHookHandler actually, not here (so we would have to change the GenericVMHookHandler interface)

Reply:
One thing that is for sure missing is initing the stateless_contracts (it should be at least the pool manager)

Reply:
huummm I see that for the EVMPoolState we are actually doing this in the builder though. meh I won't insist on where we should do this

Reply:
I'm not sure about the balances thingy (or even what they mean) but for sure we need the stateless contracts and maybe the tokens (I think it will depend on the hook)

Reply:
@dianacarvalho1 I'm a bit confused by this message... what is the point of the params if none of these attributes are needed after all? We aren't passing the params to the handler so they can't really be used there... having a hard time understanding this class 

I don't remember this conversation - without having actually started this task it's hard for me to remember relevant convos if we didn't write them down :/